src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/Tranche.java

```diff
@@ -0,0 +1,260 @@
+        this(targetTruthSensitivity, minVQSLod, numKnown, knownTiTv, numNovel, novelTiTv, accessibleTruthSites, callsAtTruthSites, model, "anonymous");
+    }
+
+    public Tranche(double targetTruthSensitivity, double minVQSLod, int numKnown, double knownTiTv, int numNovel, double novelTiTv, int accessibleTruthSites, int callsAtTruthSites, VariantRecalibratorArgumentCollection.Mode model, String name) {
+        if ( targetTruthSensitivity < 0.0 || targetTruthSensitivity > 100.0) {
+            throw new UserException("Target FDR is unreasonable " + targetTruthSensitivity);
+        }
+
+        if ( numKnown < 0 || numNovel < 0) {
+            throw new GATKException("Invalid tranche - no. variants is < 0 : known " + numKnown + " novel " + numNovel);
+        }
+
+        if ( name == null ) {
+            throw new GATKException("BUG -- name cannot be null");
+        }
+
+        this.targetTruthSensitivity = targetTruthSensitivity;
+        this.minVQSLod = minVQSLod;
+        this.novelTiTv = novelTiTv;
+        this.numNovel = numNovel;
+        this.knownTiTv = knownTiTv;
+        this.numKnown = numKnown;
+        this.model = model;
+        this.name = name;
+
+        this.accessibleTruthSites = accessibleTruthSites;
+        this.callsAtTruthSites = callsAtTruthSites;
+    }
+
+    public static final Comparator<Tranche> TRUTH_SENSITIVITY_ORDER = (tranche1, tranche2) -> Double.compare(tranche1.targetTruthSensitivity, tranche2.targetTruthSensitivity);
+
+    @Override
+    public String toString() {
+        return String.format("Tranche targetTruthSensitivity=%.2f minVQSLod=%.4f known=(%d @ %.4f) novel=(%d @ %.4f) truthSites(%d accessible, %d called), name=%s]",
+                targetTruthSensitivity, minVQSLod, numKnown, knownTiTv, numNovel, novelTiTv, accessibleTruthSites, callsAtTruthSites, name);
+    }
+
+    private static double getRequiredDouble(Map<String,String> bindings, String key ) {
+        if ( bindings.containsKey(key) ) {
+            return Double.valueOf(bindings.get(key));
+        } else  {
+            throw new UserException.MalformedFile("Malformed tranches file.  Missing required key " + key);
+        }
+    }
+
+    private static double getOptionalDouble(Map<String,String> bindings, String key, double defaultValue ) {
+        return Double.valueOf(bindings.getOrDefault(key, String.valueOf(defaultValue)));
+    }
+
+    private static int getRequiredInteger(Map<String,String> bindings, String key) {
+        if ( bindings.containsKey(key) ) {
+            return Integer.valueOf(bindings.get(key));
+        } else {
+            throw new UserException.MalformedFile("Malformed tranches file.  Missing required key " + key);
+        }
+    }
+
+    private static int getOptionalInteger(Map<String,String> bindings, String key, int defaultValue) {
+        return Integer.valueOf(bindings.getOrDefault(key, String.valueOf(defaultValue)));
```

Comment:
In these two methods as well -- throw a `UserException` upon a `NumberFormatException`